File: frontend/src/types/graph.ts
/** Economic Graph types. */

export interface GraphNode {
  node_id: string;
  node_type: string;
  label: string;
  current_state: string;
  impact_score: number;
  metadata: Record<string, unknown>;
}

export interface GraphEdge {
  source: string;
  target: string;
  relation: string;
  weight: number;
  propagation_delay_hours: number;
  description: string;
}

export interface PropagationStep {
  step: number;
  source_node: string;
  target_node: string;
  relation: string;
  impact_transmitted: number;
  resulting_state: string;
  explanation: string;
}

export interface GraphPropagationResult {
  event_id: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
  propagation_trace: PropagationStep[];
  total_steps: number;
  max_depth_reached: number;
  critical_path: string[];
  systemic_risk_score: number;
}

File: backend/app/schemas/graph.py
"""Economic Graph schemas — nodes, edges, and propagation state."""

from __future__ import annotations
from pydantic import BaseModel, Field


class GraphNode(BaseModel):
    """A node in the economic graph."""
    node_id: str
    node_type: str              # event | commodity | sector | country | infrastructure | metric
    label: str
    current_state: str          # stable | stressed | disrupted | critical
    impact_score: float = Field(ge=0.0, le=1.0, default=0.0)
    metadata: dict = Field(default_factory=dict)


class GraphEdge(BaseModel):
    """A directed edge in the economic graph."""
    source: str                 # node_id
    target: str                 # node_id
    relation: str               # affects | delays | reprices | constrains | amplifies | hedges | exports_to | exposed_to
    weight: float = Field(ge=0.0, le=1.0, default=0.5)
    propagation_delay_hours: int = 0
    description: str = ""


class PropagationStep(BaseModel):
    """A single step in the graph propagation trace."""
    step: int
    source_node: str
    target_node: str
    relation: str
    impact_transmitted: float
    resulting_state: str
    explanation: str


class GraphPropagationResult(BaseModel):
    """Full result of economic graph simulation."""
    event_id: str
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    propagation_trace: list[PropagationStep]
    total_steps: int
    max_depth_reached: int
    critical_path: list[str]    # node_ids forming the highest-impact path
    systemic_risk_score: float = Field(ge=0.0, le=1.0)
